### PR TITLE
TW-136: Change photo when create group

### PR DESCRIPTION
**Pull request by nqhhdev**
- Issue: #136 

- Resolved:

1. Delete text in app bar search when an item is selected successfully

https://github.com/linagora/twake-on-matrix/assets/99852347/60986d29-db8e-4c63-8a31-81dfcfc39f4e

2. Add loading for floating action bottom 

https://github.com/linagora/twake-on-matrix/assets/99852347/3cc80afd-ae1e-412a-895d-6981f721cca4

3. Upload avatar failed => continue creating chat

https://github.com/linagora/twake-on-matrix/assets/99852347/1d1b8b3f-0eb5-4981-97f8-74bb0cda09c6

3. Uploaded avatar => created chat

https://github.com/linagora/twake-on-matrix/assets/99852347/5bb7f367-1dd8-4c49-8b03-cdb15e41aed2

4. Cancel uploading avatar => continue creating chat

https://github.com/linagora/twake-on-matrix/assets/99852347/fcd2b3c8-d94c-402f-b2fc-c4603f792218

### CODE REVIEW

File: lib/domain/usecase/room/upload_avatar_new_group_chat_interactor.dart [hunk 1]
```diff
@@ -0,0 +1,34 @@
+
+import 'package:dartz/dartz.dart';
+import 'package:fluffychat/app_state/failure.dart';
+import 'package:fluffychat/app_state/success.dart';
+import 'package:fluffychat/domain/app_state/room/upload_avatar_new_group_chat_state.dart';
+import 'package:flutter/foundation.dart';
+import 'package:matrix/matrix.dart';
+
+class UploadAvatarNewGroupChatInteractor {
```
Comment (hoangdat) [inline lib/domain/usecase/room/upload_avatar_new_group_chat_interactor.dart:9]: should be a common for uploading media.
```suggestion
class UploadContentInteractor {
```

File: lib/pages/new_group/new_group.dart [hunk 2]
```diff
@@ -29,21 +39,25 @@ class NewGroupController extends State<NewGroup>
...
   final groupNameTextEditingController = TextEditingController();
   final groupchatInfoScrollController = ScrollController();
+  final uploadAvatarNewGroupChatNotifier = ValueNotifier<Either<Failure, Success>?>(null);
 
   final selectedContactsMapNotifier = SelectedContactsMapChangeNotifier();
   final haveGroupNameNotifier = ValueNotifier(false);
   final isEnableEEEncryptionNotifier = ValueNotifier(true);
 
+  static const maxScrollOffsetAllowedInPixel = 380.0;
+  final int maxFileSizeDefaultInMB = 5 * 1024 * 1024;
```
Comment (hoangdat) [inline lib/pages/new_group/new_group.dart:53]: really? I dont thjnk so

File: lib/pages/new_group/new_group.dart [hunk 3]
```diff
@@ -155,6 +169,102 @@ class NewGroupController extends State<NewGroup>
...
+      SheetAction(
+        key: AvatarAction.file,
+        label: L10n.of(context)!.openGallery,
+        icon: Icons.photo_outlined,
+      ),
+    ];
+
+    final action = actions.length == 1
+        ? actions.single.key
+        : await showModalActionSheet<AvatarAction>(
```
Comment (hoangdat) [inline lib/pages/new_group/new_group.dart:222]: Why the `sheetAction` look like a widget of `iOS`. How to make it more M3 style ?

File: lib/pages/new_group/new_group_chat_info.dart [hunk 4]
```diff
@@ -181,35 +216,35 @@ class NewGroupChatInfo extends StatelessWidget {
...
+  //               onSwitchButtonChanged: ((valueChanged) => {}),
+  //               isEditable: false,
+  //               defaultSwitchValue: value,
+  //             );
+  //           },
+  //         )
+  //       ],
+  //     ),
+  //   );
+  // }
```
Comment (hoangdat) [inline lib/pages/new_group/new_group_chat_info.dart:249]: delete it.

**Review by hoangdat — COMMENTED** [review]
I think it is in progress?

**Comment by sherlockvn** [comment]
> * Issue: [Change photo when create group #136](https://github.com/linagora/twake-on-matrix/issues/136)
> * Resolved:
> 
> 1. Preview avatar and get uri when selected image
> 
>  Simulator.Screen.Recording.-.iPhone.14.-.2023-07-13.at.01.34.37.mp4

Should change the border radius of avatar shadow when user click to avatar

File: lib/domain/app_state/room/upload_avatar_new_group_chat_state.dart
Comment (hoangdat) [inline lib/domain/app_state/room/upload_avatar_new_group_chat_state.dart:1]: change the name also

File: lib/domain/usecase/room/upload_avatar_new_group_chat_interactor.dart
Comment (hoangdat) [inline lib/domain/usecase/room/upload_avatar_new_group_chat_interactor.dart:1]: change the name also

**Review by hoangdat — COMMENTED** [review]
quite good for me 👍

File: lib/domain/app_state/room/upload_content_state.dart [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  final Uri uri;
+  final Uint8List file;
+
+  const UploadContentSuccess({
+    required this.uri,
+    required this.file
+  });
+
+  @override
+  List<Object?> get props => [uri];
```
Comment (sherlockvn) [inline lib/domain/app_state/room/upload_content_state.dart:22]: missing file

**Comment by sherlockvn** [comment]
> * Issue: [Change photo when create group #136](https://github.com/linagora/twake-on-matrix/issues/136)
> * Resolved:
> 
> 1. Delete text in app bar search when an item is selected successfully
> 
>  Screen.Recording.2023-07-17.at.03.52.43.mov 
> 2. Add loading for floating action bottom
> 
>  Screen.Recording.2023-07-14.at.14.25.13.mov 
> 3. Upload avatar failed => continue creating chat
> 
>  Screen.Recording.2023-07-14.at.01.57.10.mov 
> 3. Uploaded avatar => created chat
> 
>  Screen.Recording.2023-07-14.at.02.03.20.mov 
> 4. Cancel uploading avatar => continue creating chat
> 
>  Screen.Recording.2023-07-14.at.02.30.58.mov

Avatar missing when click to fab button, i think it should not

File: lib/domain/usecase/room/upload_content_interactor.dart [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+          yield Left(FileTooBigMatrix(FileTooBigMatrixException(fileSize, maxMediaSize)));
+        }
+        final uri = await matrixClient.uploadContent(
+          matrixFile.bytes,
+          filename: matrixFile.name,
+          contentType: matrixFile.mimeType,
+        );
+        yield Right(UploadContentSuccess(uri: uri, file: matrixFile.bytes));
+      } else {
+        yield const Left(UploadContentFailed(exception: null));
```
Comment (imGok) [inline lib/domain/usecase/room/upload_content_interactor.dart:33]: what about making exception optional in UploadContentFailed so you don't have to initialize it to null

File: lib/pages/new_group/new_group.dart [hunk 7]
```diff
@@ -155,6 +176,186 @@ class NewGroupController extends State<NewGroup>
...
+  }
+
+  void imagePickAction() async {
+    Navigator.pop(context);
+    final assetEntity = await CameraPicker.pickFromCamera(
+      context,
+      locale: window.locale,
+    );
+    if (assetEntity != null && assetEntity.type == AssetType.image) {
+      uploadAvatarNewGroupChat(
```
Comment (imGok) [inline lib/pages/new_group/new_group.dart:277]: Why should we upload the avatar on image pick ? We might upload unecessary images if user picks image several time

Comment (imGok) [inline lib/pages/new_group/new_group.dart:277]: Why should we upload the image everytime the user picks a file ? At the end we might upload unused images, if the user changed his mind

Comment (hoangdat) [reply lib/pages/new_group/new_group.dart:277]: good idea to delete the unused avatar

File: lib/domain/usecase/room/upload_content_interactor.dart (hunk 6 above)
Comment (nqhhdev) [reply lib/domain/usecase/room/upload_content_interactor.dart]: Yes, I update `CannotUploadContentException()` in here

File: lib/pages/new_group/new_group.dart (hunk 7 above)
Comment (nqhhdev) [reply lib/pages/new_group/new_group.dart:256]: We will implement this case on this ticket #318

File: lib/domain/app_state/room/upload_content_state.dart [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import 'package:matrix/matrix.dart';
+
+class UploadContentLoading extends Success {
+  @override
+  List<Object?> get props => [];
+}
+
+class UploadContentSuccess extends Success {
+  final Uri uri;
+  final Uint8List file;
```
Comment (hoangdat) [inline lib/domain/app_state/room/upload_content_state.dart:14]: why we need to respond `file` as `bytes` here?

File: lib/domain/exception/room/can_not_create_new_group_chat_exception.dart [hunk 9]
```diff
@@ -0,0 +1,3 @@
+class CannotCreateNewGroupChatException implements Exception {
+  CannotCreateNewGroupChatException() : super();
```
Comment (hoangdat) [inline lib/domain/exception/room/can_not_create_new_group_chat_exception.dart:2]: how about the root cause `e`? Please don't swallow the error

File: lib/pages/new_private_chat/search_contacts_controller.dart [hunk 10]
```diff
@@ -18,18 +18,16 @@ class SearchContactsController {
   final TextEditingController textEditingController = TextEditingController();
   StreamController<Either<Failure, GetContactsSuccess>> lookupStreamController = StreamController();
   void Function(String)? onSearchKeywordChanged;
-  late final isSearchModeNotifier = ValueNotifier(false);
+  ValueNotifier<bool> isSearchModeNotifier = ValueNotifier(false);
+  ValueNotifier<bool> isSearchBarShow = ValueNotifier(false);
```
Comment (hoangdat) [inline lib/pages/new_private_chat/search_contacts_controller.dart:22]: why we need 2 variables for it? why not use only one?

File: lib/pages/new_group/new_group_info_controller.dart [hunk 11]
```diff
@@ -1,42 +1,67 @@
...
+          onAcceptButton: () => cancelUploadAvatar(),
+        ),
+      );
     }
   }
 
+  void cancelUploadAvatar() {
+    uploadAvatarNewGroupChatNotifier.value = const Left(UploadContentFailed(exception: null));
+    removeAllImageSelected();
+    Navigator.pop(context);
```
Comment (hoangdat) [inline lib/pages/new_group/new_group_info_controller.dart:60]: why not use `vroute`?

Comment (hoangdat) [reply lib/pages/new_group/new_group_info_controller.dart:60]: and why not pop when cancel uploading?

File: lib/pages/new_group/new_group.dart [hunk 12]
```diff
@@ -131,6 +155,186 @@ class NewGroupController extends State<NewGroup>
...
+        matrixClient: Matrix.of(context).client,
+        entity: imagePickerController.selectedAssets.first.asset
+      );
+    }
+  }
+
+  void _registerListenerForSelectedImagesChanged() {
+    imagePickerController.addListener(() {
```
Comment (hoangdat) [inline lib/pages/new_group/new_group.dart:264]: please make sure, user can not select more than one image in this case

**Comment by hoangdat** [comment]
@nqhhdev Khải is trying to separate `Image_picker_mixin`, can we can reuse it?

File: lib/pages/new_group/new_group.dart (hunk 12 above)
Comment (nqhhdev) [reply lib/pages/new_group/new_group.dart:264]: if (numberSelectedImagesNotifier.value == 1)

Yes, make sureee

Comment (hoangdat) [reply lib/pages/new_group/new_group.dart:264]: I mean, in the screen, what happen if user select more than one.

File: lib/pages/new_group/new_group.dart (hunk 2 above)
Comment (hoangdat) [reply lib/pages/new_group/new_group.dart:61]: use `FutureBuilder` to show media config

**Comment by hoangdat** [comment]
- clear text and keep focus after search and select a contact
- empty group name: disable create button
- group name input: 
   - if members <= 3: pick display name: `A, B and C` and select all text, focus in here, show keyboard in mobile
   - else: group name empty

**Comment by nqhhdev** [comment]
> * clear text and keep focus after search and select a contact
> * if empty group name: disable create button
> * group name input:
>   
>   * if members <= 3: pick display name: `A, B and C` and select all text, focus in here, show keyboard in mobile
>   * else: group name empty


My demo


https://github.com/linagora/twake-on-matrix/assets/99852347/d3f31b3a-9c9d-40eb-989f-89c71cd103ee

File: lib/pages/new_group/new_group.dart [hunk 13]
```diff
@@ -119,22 +116,47 @@ class NewGroupController extends State<NewGroup>
...
+        displayName: isCustomDisplayName ? L10n.of(context)!.you : profile.displayname,
         matrixId: Matrix.of(context).client.userID,
       )
     };
     newContactsList.addAll(getSelectedValidContacts(contactsList));
     return newContactsList;
   }
 
+  void _getDefaultGroupName() async {
+    final newContactsList = await getAllContactsGroupChat(isCustomDisplayName: false);
```
[the remark this answers is not in the document]
Comment (nqhhdev) [reply lib/pages/new_group/new_group.dart:146]: 1. This is the list after being selected on the screen I create new group.
2. You are wrong, I am not calling API here, I think being called twice, I will reuse it

File: lib/config/routes.dart [hunk 14]
```diff
@@ -120,9 +120,9 @@ class AppRoutes {
                   stackedRoutes: _chatDetailsRoutes,
                   buildTransition: rightToLeftTransition,
                 ),
-                VWidgetWithDependency(
+                VWidgetWithDependencies(
                   path: '/rooms/:roomid',
-                  di: ChatScreenDi(),
+                  dIs: [ChatScreenDi(), CreateNewRoomDI()],
```
Comment (hoangdat) [inline lib/config/routes.dart:125]: and here

**Review by hoangdat — APPROVED** [review]

**Pull request merged**